fuse/utils/interface_validator.py

```diff
@@ -0,0 +1,156 @@
+from typing import List, Dict, Callable, Any, _UnionGenericAlias
+
+
+import inspect
+
+# https://stackoverflow.com/questions/147816/preserving-signatures-of-decorated-functions
+# https://stackoverflow.com/questions/5929107/decorators-with-parameters
+# https://stackoverflow.com/questions/2366713/can-a-decorator-of-an-instance-method-access-the-class
+
+
+"""
+This is a very handy class that allows enforcing an API
+Unlike programming languages like C++, python does not have a built in way to enforce adherance to API when inheriting from a base class
+By using the code found in this script you can require the inheriting class to implement certain method and also control which arguments of the method signature is required.
+an actively used example can be seen in InsilicoLabModelBase() in insilico-lab-validation repo.
+For illustration, you can create a base class FruitBase (you can see related unit test in `fuse-med-ml/fuse/utils/tests/test_interface_validator.py`)
+
+
+```python
+from fuse.utils.interface_validator import InterfaceValidator, validate_signature
+
+class FruitBase(InterfaceValidator):
+    def __init__(self, *, wow='wow'):
+        super().__init__()
+        self.wow = wow
+
+    @validate_signature(args=['color'])  #only argument `color` will be validated to both exist and have the same type hint in the inheriting class
+    def draw(self, *, color:str = 'red', style:int=12, **kwargs):
+        print(f'FruitBase.draw: {color=} {style=}')
+
+    @validate_signature(args=['size', 'material']) #both args `size` and `material` will be validated
+    def print_3d(self, *, size:float = 2.7, material:str = 'iron'):
+        print(f'FruitBase.print_3d {size=} {material=}')
+```
+
+now, when you use a class that inherits from it, like:
+
+```python
+class Banana(FruitBase):
+    #...
+```
+
+the API will be enforced and exceptions will be raised if requirements are not met.
+
+
+#TODO:
+    1. allow validating return type hint
+    2. allow using @validate_signature without using any args, and that would mean that the exact same full signature is required
+    3. allow control over whether type hint is validated or not (is this needed?)
+"""
```

Comment:
Also validating that there are no additional arguments without default values is important

Reply:
can you elaborate, I don't think I understand what you meant